### ISSUE TITLE
Failed to build afopencl x64 target on MSVS 2017

### ISSUE THREAD

**Issue by laser101** [arrayfire/arrayfire#2924]
Hi,

Building `afopencl` target on x64 MSVS 2017 is failing at master d2df83d (5 June 2020) with a bunch of errors relating to "declaration is not allowed in 'constexpr' function body". The culprit appears to be `getTypeBuildDefinition()` defined in `\arrayfire\src\backend\opencl\types.hpp:128`.

Full error message...

```
10>git\arrayfire\src\backend\opencl\types.hpp(137): error C3546: '...': there are no parameter packs available to expand
10>E:\Russ\SwDev\lu_dev\git\arrayfire\src\backend\opencl\kernel/reduce.hpp(139): note: see reference to function template instantiation 'const char *opencl::getTypeBuildDefinition<Ti,To>(void)' being compiled
10>        with
10>        [
10>            Ti=float,
10>            To=char
10>        ]
10>E:\Russ\SwDev\lu_dev\git\arrayfire\src\backend\opencl\kernel/reduce.hpp(232): note: see reference to function template instantiation 'void opencl::kernel::reduceFirstLauncher<Ti,To,af_and_t>(opencl::Param,opencl::Param,const opencl::uint,const opencl::uint,const opencl::uint,int,double)' being compiled
10>        with
10>        [
10>            Ti=float,
10>            To=char
10>        ]
10>git\arrayfire\src\backend\opencl\types.hpp(135): error C3250: 'is_half': declaration is not allowed in 'constexpr' function body
10>git\arrayfire\src\backend\opencl\types.hpp(136): error C3250: 'is_double': declaration is not allowed in 'constexpr' function body
10>git\arrayfire\src\backend\opencl\types.hpp(139): error C3250: 'half_def': declaration is not allowed in 'constexpr' function body
10>git\arrayfire\src\backend\opencl\types.hpp(141): error C3250: 'double_def': declaration is not allowed in 'constexpr' function body
10>git\arrayfire\src\backend\opencl\types.hpp(128): error C3259: 'constexpr' functions can only have one return statement
```

Russ

**Comment by 9prady9** [comment]
@laser101 Interesting, we have moved to C++14 since a few months. I wonder if this could be because of MSVC doesn't have full support in the toolset version you are using.

Can you please try the following patch and let me know if helps.
```diff
diff --git a/src/backend/opencl/types.hpp b/src/backend/opencl/types.hpp
index 83a5d624c..f06f889d3 100644
--- a/src/backend/opencl/types.hpp
+++ b/src/backend/opencl/types.hpp
@@ -132,13 +132,13 @@ constexpr const char *getTypeBuildDefinition() {
     using std::begin;
     using std::end;
     using std::is_same;
-    array<bool, sizeof...(ARGS)> is_half = {is_same<ARGS, half>::value...};
-    array<bool, sizeof...(ARGS) * 2> is_double = {
+    constexpr array<bool, sizeof...(ARGS)> is_half = {is_same<ARGS, half>::value...};
+    constexpr array<bool, sizeof...(ARGS) * 2> is_double = {
         is_same<ARGS, double>::value..., is_same<ARGS, cdouble>::value...};
 
-    bool half_def =
+    constexpr bool half_def =
         any_of(begin(is_half), end(is_half), [](bool val) { return val; });
-    bool double_def =
+    constexpr bool double_def =
         any_of(begin(is_double), end(is_double), [](bool val) { return val; });
 
     if (half_def && double_def) {
```

**Comment by laser101** [comment]
I'll review the patch and get back to you.

Could you just confirm that I am calling cmake correctly for my MSVS 2017 environment please, just in case I'm introducing a problem here.
```
cmake -Tv140 -G "Visual Studio 15 2017 Win64" -DCMAKE_TOOLCHAIN_FILE="\git\vpkg\scripts\buildsystems\vcpkg.cmake" ..
```

Thanks,

Russ

**Comment by 9prady9** [comment]
> I'll review the patch and get back to you.
> 
> Could you just confirm that I am calling cmake correctly for my MSVS 2017 environment please, just in case I'm introducing a problem here.
> ```
> cmake -Tv140 -G "Visual Studio 15 2017 Win64" -DCMAKE_TOOLCHAIN_FILE="\git\vpkg\scripts\buildsystems\vcpkg.cmake" ..
> ```
> 
> Thanks,
> 
> Russ

please make sure if there is a space needed between T and it's value. Otherwise, looks good to me.

Either way, the error you posted earlier couldn't possibly come with cmake error I would think.

**Comment by laser101** [comment]
The `cmake` command line was taken from the [build instructions](https://github.com/arrayfire/arrayfire/wiki/Build-Instructions-for-Windows#arrayfire-build-instructions) where the `-Tv140` option is included without a space after the `-T`. Does the documentation need updating?

I'll provide an update on the patch soon.

**Comment by 9prady9** [comment]
> The `cmake` command line was taken from the [build instructions](https://github.com/arrayfire/arrayfire/wiki/Build-Instructions-for-Windows#arrayfire-build-instructions) where the `-Tv140` option is included without a space after the `-T`. Does the documentation need updating?
> 
> I'll provide an update on the patch soon.

I was merely being cautious as I don't remember the syntax of the top of my head. If you did face issues because of that earlier. Sure, then it needs to be fixed in the documentation. Please let me know if that was the case earlier.

**Comment by 9prady9** [comment]
@laser101 I just realized couple of things w.r.t C++14 support. Though your cmake generator is `VS 2017` the toolset you are using v140 which is basically from VS 2015 update. Hence, the `constexpr` support isn't complete. Extended `constexpr` is available in VS from 2017 and VS 2017 is C++14 feature complete since that version only.

I don't think this can be fixed without code change as extended constexpr support got added from v141 (VS 2017).

Can you try with `v141` toolset ? It should be by default installed with VS 2017 if `Desktop C++ Development` component was installed.

**Comment by 9prady9** [comment]
I personally think that we should raise the minimum required VS to VS 2017 (v141). @umar456 what's your thought on that ?

**Comment by 9prady9** [comment]
@laser101 Kindly ignore the earlier patch, and please try this one https://github.com/arrayfire/arrayfire/pull/2925

You can fetch the patch file from [here](https://github.com/arrayfire/arrayfire/pull/2925.patch)

**Comment by 9prady9** [comment]
@laser101 Kindly ignore the earlier patch, and please try this one https://github.com/arrayfire/arrayfire/pull/2925

You can fetch the patch file from [here](https://github.com/arrayfire/arrayfire/pull/2925.patch)

**Comment by laser101** [comment]
Thanks for your comments re. the MS Visual C++ toolset. I must admit that I wasn't very clear what the `cmake -T140` option was doing, but copied it verbatum from the [build instructions](https://github.com/arrayfire/arrayfire/wiki/Build-Instructions-for-Windows#arrayfire-build-instructions) which suggests to use this setting for "Visual Studio 2017 v15.5 and above".

Following your suggestion, and with a bit of [background reading](https://en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B#Internal_version_numbering), I now understand that I should be using `cmake -T141`. Indeed, master branch d2df83d (5 June 2020) now builds fine without any additional patching if I use toolset 141.

So please accept my apologies for posting issues that now seem specific to building with toolset 140, when in fact I should be building with toolset 141. Can I suggest that the build instructions are updated to reflect this as I believe they are currently a bit misleading.

As a side note, I was able to build release 3.7.0 with MSVS 2017 toolset 140 - perhaps that is the last release that is compatible with that toolset. Perhaps that could also be added to the build instructions?

I think this issue can now be closed.

**Comment by 9prady9** [comment]
> @laser101 Kindly ignore the earlier patch, and please try this one https://github.com/arrayfire/arrayfire/pull/2925
> 
> You can fetch the patch file from [here](https://github.com/arrayfire/arrayfire/pull/2925.patch)

Can you please test with #2925 PR changes - it should handle the build with Tv140 toolset also.

**Comment by laser101** [comment]
Hi,

Sorry I'm new to GitHub pull requests - how do I download a diff for the 5 changed files? I can browse the changed files in commit 62f6ced and I can download the entire branch at commit 62f6ced as a ZIP file, but I don't see a away to download a diff.

When I download the ZIP file and run the following, `cmake` is failing.

```
cmake -T v140 -G "Visual Studio 15 2017 Win64" -DCMAKE_TOOLCHAIN_FILE="git\vpkg\scripts\buildsystems\vcpkg.cmake" ..
```

```
CMake Error: File /git/arrayfire-62f6ced8da86b1232c661f0c53765494e1837bbc/extern/forge/CMakeModules/version.h.in does not exist.
CMake Error at CMakeModules/AFconfigure_forge_submodule.cmake:47 (configure_file):
  configure_file Problem configuring file
Call Stack (most recent call first):
  CMakeLists.txt:117 (include)


CMake Error at CMakeLists.txt:163 (add_subdirectory):
  The source directory

    /git/arrayfire-62f6ced8da86b1232c661f0c53765494e1837bbc/extern/spdlog

  does not contain a CMakeLists.txt file.


CMake Error at CMakeLists.txt:164 (add_subdirectory):
  The source directory

    /git/arrayfire-62f6ced8da86b1232c661f0c53765494e1837bbc/extern/glad

  does not contain a CMakeLists.txt file.


-- CUDA_architecture_build_targets: 7.0 6.1
-- Found OpenCL: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v9.2/lib/x64/OpenCL.lib (found version "1.2")
CMake Error at test/CMakeLists.txt:27 (add_subdirectory):
  The source directory

    /git/arrayfire-62f6ced8da86b1232c661f0c53765494e1837bbc/test/gtest

  does not contain a CMakeLists.txt file.


CMake Error at test/CMakeLists.txt:28 (set_target_properties):
  set_target_properties Can not find target to add properties to: gtest


-- Performing Test pthread_flag
-- Performing Test pthread_flag - Failed
fatal: not a git repository (or any of the parent directories): .git
-- No git. Setting hash to default
-- Configuring incomplete, errors occurred!
See also "/git/arrayfire-62f6ced8da86b1232c661f0c53765494e1837bbc/build/CMakeFiles/CMakeOutput.log".
See also "/git/arrayfire-62f6ced8da86b1232c661f0c53765494e1837bbc/build/CMakeFiles/CMakeError.log".
```

I assume that I am just trying to use the pull request incorrectly, so any tips would be appreciated.

Russ

**Comment by 9prady9** [comment]
That is why I shared the patch URL as hyperlink in my earlier comments.
Just in case, it is posted below again

https://github.com/arrayfire/arrayfire/pull/2925.patch

1.Open the link in your browser.
2. Save the file to disk.
3. now run the command "git apply patch_file-name"

Now run the build command as usual.

**Comment by laser101** [comment]
Sorry, I just spotted that before you posted. I'm evaluating it now.

**Comment by laser101** [comment]
Ok, having applied the patch and attempted to build with toolset 140, I am still getting errors related to `constexpr`. I've attached the build error log here... 
[build-errors.txt](https://github.com/arrayfire/arrayfire/files/4745644/build-errors.txt)

Regards,

Russ

**Comment by 9prady9** [comment]
@laser101 thank you for testing it out. I will look into it.

**Comment by 9prady9** [comment]
I was able to reproduce it, I am working on the fix and will send in the patch to #2925 soon.
